core/src/test/java/io/grpc/internal/ManagedChannelServiceConfigTest.java

```diff
@@ -215,6 +216,84 @@ public void getDefaultConfigSelectorFromConfig() {
         .isEqualTo(serviceConfig.getMethodConfig(method));
   }
 
+  @Test
+  public void retryConfig_emptyRetriableStatusCodesAllowedWithPerAttemptRecvTimeoutGiven() {
+    Map<String, ?> retryPolicy = ImmutableMap.<String, Object>builder()
+        .put("maxAttempts", 3.0D)
+        .put("initialBackoff", "1s")
+        .put("maxBackoff", "10s")
+        .put("backoffMultiplier", 1.5D)
+        .put("perAttemptRecvTimeout", "2.5s")
+        .put("retryableStatusCodes", ImmutableList.of())
+        .build();
+    Map<String, ?> methodConfig = ImmutableMap.of(
+        "name", ImmutableList.of(ImmutableMap.of()), "retryPolicy", retryPolicy);
+    Map<String, ?> rawServiceConfig =
+        ImmutableMap.of("methodConfig", ImmutableList.of(methodConfig));
+    assertThat(ManagedChannelServiceConfig.fromServiceConfig(rawServiceConfig, true, 5, 5, null))
+        .isNotNull();
+  }
+
+  @Test
+  public void retryConfig_PerAttemptRecvTimeoutUnsetAllowedIfRetryableStatusCodesNonempty() {
+    Map<String, ?> retryPolicy = ImmutableMap.<String, Object>builder()
+        .put("maxAttempts", 3.0D)
+        .put("initialBackoff", "1s")
+        .put("maxBackoff", "10s")
+        .put("backoffMultiplier", 1.5D)
+        .put("retryableStatusCodes", ImmutableList.of("UNAVAILABLE"))
+        .build();
+    Map<String, ?> methodConfig = ImmutableMap.of(
+        "name", ImmutableList.of(ImmutableMap.of()), "retryPolicy", retryPolicy);
+    Map<String, ?> rawServiceConfig =
+        ImmutableMap.of("methodConfig", ImmutableList.of(methodConfig));
+    assertThat(ManagedChannelServiceConfig.fromServiceConfig(rawServiceConfig, true, 5, 5, null))
+        .isNotNull();
+  }
+
+  @Test
+  public void retryConfig_emptyRetriableStatusCodesNotAllowedWithPerAttemptRecvTimeoutUnset() {
+    Map<String, ?> retryPolicy = ImmutableMap.<String, Object>builder()
+        .put("maxAttempts", 3.0D)
+        .put("initialBackoff", "1s")
+        .put("maxBackoff", "10s")
+        .put("backoffMultiplier", 1.5D)
+        .put("retryableStatusCodes", ImmutableList.of())
+        .build();
+    Map<String, ?> methodConfig = ImmutableMap.of(
+        "name", ImmutableList.of(ImmutableMap.of()), "retryPolicy", retryPolicy);
+    Map<String, ?> rawServiceConfig =
+        ImmutableMap.of("methodConfig", ImmutableList.of(methodConfig));
+    try {
+      ManagedChannelServiceConfig.fromServiceConfig(rawServiceConfig, true, 5, 5, null);
+      fail("The expected IllegalArgumentException is not thrown");
+    } catch (IllegalArgumentException e) {
+      assertThat(e).hasMessageThat().contains(
+          "retryableStatusCodes cannot be empty without perAttemptRecvTimeout");
+    }
+  }
+
+  // For now we allow perAttemptRecvTimeout being zero although it does not make sense.
+  // TODO(zdapeng): disallow zero perAttemptRecvTimeout if hedging is not enabled once we support
```

Comment:
Once retries are no longer experimental, we can't be adding a restriction. Right? And it sounded like your plan was to try to get retries stable here before supporting hedging.

Reply:
> Once retries are no longer experimental, we can't be adding a restriction. Right?

That's right.

> And it sounded like your plan was to try to get retries stable here before supporting hedging.

Why? I'm not.

I'm currently allowing zero perAttemptRecvTimeout, because in the future, we have usecase for hedging.
Once hedge_on_per_try_timeout is supported, we will add restriction for zero perAttemptRecvTimeout without hedge_on_per_try_timeout=true. The planned change is still before getting retries stable.

Reply:
Oh, I see.